Feature/stlnd 1924

MaciekKrz opened the pull request:
## Ticket

https://stella-systems.atlassian.net/browse/STLND-1924

## Summary

- Fixed MQTT Sink Reconnection Bug: Resolved an issue where the SDK failed to handle message sending when the MQTT broker connection was lost. The SDK now gracefully manages reconnection and resends queued messages, ensuring reliable delivery.

- Enhanced on_disconnect Handler: Added a robust handler for the MQTT sink’s on_disconnect callback to better manage disconnection events and trigger reconnection logic.

- Enabled Strict mypy Checks: Integrated strict type checking with mypy --strict --check-untyped-defs, catching and resolving type-related issues across the codebase.

- Improved Static Typing: Refined type annotations for better code safety and maintainability, including precise types for MQTT callbacks, message queues, and authentication tasks.


## Test Plan

Tested by PG

## Documentation

N/A

MaciekKrz committed:
[STLND1924] Fixed message queue behavoiur during broker reconnections

- added strict checks for mypy
- improved static types
- added proper handler for mqtt sink `on_disconnect`

MaciekKrz committed:
[STLND-1924] Static typing improvements

MaciekKrz committed:
[STLND-1924] Version upgrade

sebtom commented on stellanow_sdk_python/sinks/mqtt/stellanow_mqtt_sink.py:
```diff
@@ -50,131 +47,140 @@ def __init__(
-        logger.error(f"Handling publish error: {error}")
-        await self.wait_for_connection()
-        logger.info("Retrying message publish...")
+                await asyncio.sleep(2.5)
```
2,5 seconds sleep? how we get to this number? :) so every 2,5s we check if connection is up?

what will happen if during that time someone call `send_message` and it fails with `raise Exception("MQTT sink is not connected.")`

should the `send` method reconnect maybe?

MaciekKrz replied:
I picked 2.5s as a balance between responsiveness and resource usage, also I've just checked the TS SDK and it seems that there is the same time set for connection checks. 

If during the `send_message` there is any connection issue, messages are being `enqueue` in the queue and waiting for reconnection. 
The reconnection itself is handled by the `_connection_monitor`.
I can improve the exception and logger in the `send_message` method:
```
if not self.is_connected():
            logger.warning(f"Cannot send message {message.message_id}: MQTT sink is disconnected. Awaiting reconnection...")
            raise Exception("MQTT sink is disconnected; connection monitor is attempting to reconnect.")
```

sebtom replied:
maybe we should put the reconnection time as parameter when we initialize SDK by the user?

I can see now you added in "send"
<img width="627" alt="image" src="https://github.com/user-attachments/assets/75b2a550-5b0a-4c3d-ae2d-7a123d45f843" />

MaciekKrz replied:
It's not a good idea to make it configurable. If someone sets it too high, the SDK will not work optimally.

MaciekKrz commented on stellanow_sdk_python/sdk.py:
```diff
@@ -57,14 +58,18 @@ def __init__(self, project_info: StellaProjectInfo, sink: IStellaNowSink, queue_
         """
-        await self.sink.connect()
+        await self.sink.connect()  # Blocks until connected
+        while not self.sink.is_connected():
```
That while loop was introduced due to line
`await asyncio.wait_for(self._is_connected_event.wait(), timeout=30)` in connect method from StellaNowMqttSink.
So it blocks that connection to sink but after 30 seconds there was a timeout and an additional check was required.

But I've tested it with `timeout=None` and it seems to work fine. In such a scenario, the while loop is not needed anymore.

MaciekKrz committed:
[STLND-1924] Improved reconnection to broker

- Improved reconnection to broker by setting timeout=None in `connect` method
- Updated execptiom description for sink `send_message` method
- Removed while loop in sdk `start` method

MaciekKrz committed:
[STLND-1924] Lint

sebtom reviewed (commented):
approving to not block with minor comments
